File: frontend/src/lib/api/verifier.ts
import "server-only";
import type { ClaimRecord } from "@/lib/api/claims";
import { gatewayGet, gatewayPost } from "@/lib/api/gateway";

export type DecisionOutcome =
  "approved" | "rejected" | "more_information_requested";

type ApiAIReview = {
  assessment: string;
  confidence: string | null;
  extracted_figures: Record<string, string> | null;
  flags: string[] | null;
  narrative: string;
  assessed_by: string;
  assessed_at: string;
};

type ApiDecision = {
  id: string;
  verifier_user_id: string;
  verifier_name: string;
  outcome: DecisionOutcome;
  approved_amount: string | null;
  reason: string;
  decided_at: string;
};

export type AIReviewRecord = {
  assessment: string;
  confidence: string | null;
  extractedFigures: Record<string, string>;
  flags: string[];
  narrative: string;
  assessedBy: string;
  assessedAt: string;
};

export type DecisionRecord = {
  id: string;
  verifierUserId: string;
  verifierName: string;
  outcome: DecisionOutcome;
  approvedAmount: string | null;
  reason: string;
  decidedAt: string;
};

export type ReviewRecord = {
  claim: ClaimRecord;
  aiReview: AIReviewRecord | null;
  decisions: DecisionRecord[];
};

const toAIReview = (review: ApiAIReview | null): AIReviewRecord | null =>
  review
    ? {
        assessment: review.assessment,
        confidence: review.confidence,
        extractedFigures: review.extracted_figures ?? {},
        flags: review.flags ?? [],
        narrative: review.narrative,
        assessedBy: review.assessed_by,
        assessedAt: review.assessed_at,
      }
    : null;

const toDecision = (decision: ApiDecision): DecisionRecord => ({
  id: decision.id,
  verifierUserId: decision.verifier_user_id,
  verifierName: decision.verifier_name,
  outcome: decision.outcome,
  approvedAmount: decision.approved_amount,
  reason: decision.reason,
  decidedAt: decision.decided_at,
});

export const fetchReviewQueue = async (
  token: string,
): Promise<ClaimRecord[]> => {
  const { fetchClaimsFromPayload } = await import("@/lib/api/claims");
  const queued = await gatewayGet<{ claims: unknown[] }>(
    "/v1/verifier/queue",
    token,
  );
  return fetchClaimsFromPayload(queued.claims);
};

export const fetchReview = async (
  token: string,
  claimId: string,
): Promise<ReviewRecord> => {
  const { fetchClaimsFromPayload } = await import("@/lib/api/claims");
  const reviewed = await gatewayGet<{
    claim: unknown;
    ai_review: ApiAIReview | null;
    decisions: ApiDecision[];
  }>(`/v1/verifier/claims/${claimId}`, token);

  return {
    claim: fetchClaimsFromPayload([reviewed.claim])[0]!,
    aiReview: toAIReview(reviewed.ai_review),
    decisions: reviewed.decisions.map(toDecision),
  };
};

export type DecisionDraft = {
  outcome: DecisionOutcome;
  approvedAmount?: string;
  reason?: string;
};

export const decideClaim = async (
  token: string,
  claimId: string,
  draft: DecisionDraft,
  idempotencyKey: string,
): Promise<ReviewRecord> => {
  const { fetchClaimsFromPayload } = await import("@/lib/api/claims");
  const decided = await gatewayPost<{
    claim: unknown;
    decisions: ApiDecision[];
  }>(
    `/v1/verifier/claims/${claimId}/decision`,
    token,
    {
      outcome: draft.outcome,
      approved_amount: draft.approvedAmount ?? "",
      reason: draft.reason ?? "",
    },
    idempotencyKey,
  );

  return {
    claim: fetchClaimsFromPayload([decided.claim])[0]!,
    aiReview: null,
    decisions: decided.decisions.map(toDecision),
  };
};

File: backend/services/api-gateway/internal/handler/verifier.go
package handler

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"

	sustainabilityv1 "github.com/carboncircuit/backend/gen/carboncircuit/sustainability/v1"
	"github.com/carboncircuit/backend/internal/httpx"
	"github.com/carboncircuit/backend/services/api-gateway/internal/caller"
)

var decisionOutcomeByName = map[string]sustainabilityv1.DecisionOutcome{
	"approved":                   sustainabilityv1.DecisionOutcome_DECISION_OUTCOME_APPROVED,
	"rejected":                   sustainabilityv1.DecisionOutcome_DECISION_OUTCOME_REJECTED,
	"more_information_requested": sustainabilityv1.DecisionOutcome_DECISION_OUTCOME_MORE_INFORMATION_REQUESTED,
}

var decisionOutcomeName = invertMap(decisionOutcomeByName)

type decideRequest struct {
	Outcome        string `json:"outcome" binding:"required"`
	ApprovedAmount string `json:"approved_amount" binding:"max=32"`
	Reason         string `json:"reason" binding:"max=2000"`
}

type aiReviewResponse struct {
	Assessment       string            `json:"assessment"`
	Confidence       *string           `json:"confidence"`
	ExtractedFigures map[string]string `json:"extracted_figures"`
	Flags            []string          `json:"flags"`
	Narrative        string            `json:"narrative"`
	AssessedBy       string            `json:"assessed_by"`
	AssessedAt       string            `json:"assessed_at"`
}

type decisionResponse struct {
	ID             string  `json:"id"`
	VerifierUserID string  `json:"verifier_user_id"`
	VerifierName   string  `json:"verifier_name"`
	Outcome        string  `json:"outcome"`
	ApprovedAmount *string `json:"approved_amount"`
	Reason         string  `json:"reason"`
	DecidedAt      string  `json:"decided_at"`
}

func toAIReviewResponse(review *sustainabilityv1.AIReview) *aiReviewResponse {
	if review == nil {
		return nil
	}

	return &aiReviewResponse{
		Assessment:       review.GetAssessment(),
		Confidence:       emptyToNil(review.GetConfidence()),
		ExtractedFigures: review.GetExtractedFigures(),
		Flags:            review.GetFlags(),
		Narrative:        review.GetNarrative(),
		AssessedBy:       review.GetAssessedBy(),
		AssessedAt:       review.GetAssessedAt(),
	}
}

func toDecisionResponses(decisions []*sustainabilityv1.ClaimDecision) []decisionResponse {
	recorded := make([]decisionResponse, 0, len(decisions))

	for _, decision := range decisions {
		recorded = append(recorded, decisionResponse{
			ID:             decision.GetId(),
			VerifierUserID: decision.GetVerifierUserId(),
			VerifierName:   decision.GetVerifierName(),
			Outcome:        decisionOutcomeName[decision.GetOutcome()],
			ApprovedAmount: emptyToNil(decision.GetApprovedAmount()),
			Reason:         decision.GetReason(),
			DecidedAt:      decision.GetDecidedAt(),
		})
	}

	return recorded
}

func (h *Handlers) ReviewQueue(c *gin.Context) {
	if _, resolved := caller.ContextFrom(c.Request.Context()); !resolved {
		httpx.Fail(c, httpx.CodeUnauthenticated)
		return
	}

	limit := 0
	if raw := c.Query("limit"); raw != "" {
		parsed, err := strconv.Atoi(raw)
		if err != nil || parsed < 1 {
			httpx.Fail(c, httpx.CodeValidation, httpx.FieldError{
				Field: "limit", Code: "UNSUPPORTED_VALUE",
			})
			return
		}
		limit = parsed
	}

	queued, err := h.Sustainability.ReviewQueue(c.Request.Context(),
		&sustainabilityv1.ReviewQueueRequest{After: c.Query("after"), Limit: int32(limit)})
	if err != nil {
		h.failClaim(c, err)
		return
	}

	claims := make([]claimResponse, 0, len(queued.GetClaims()))
	for _, claim := range queued.GetClaims() {
		claims = append(claims, toClaimResponse(claim))
	}

	httpx.Data(c, http.StatusOK, map[string]any{
		"claims":      claims,
		"next_cursor": emptyToNil(queued.GetNextCursor()),
	})
}

func (h *Handlers) ReviewClaim(c *gin.Context) {
	if _, resolved := caller.ContextFrom(c.Request.Context()); !resolved {
		httpx.Fail(c, httpx.CodeUnauthenticated)
		return
	}

	reviewed, err := h.Sustainability.ReviewClaim(c.Request.Context(), c.Param("claimId"))
	if err != nil {
		h.failClaim(c, err)
		return
	}

	httpx.Data(c, http.StatusOK, map[string]any{
		"claim":     toClaimResponse(reviewed.GetClaim()),
		"ai_review": toAIReviewResponse(reviewed.GetAiReview()),
		"decisions": toDecisionResponses(reviewed.GetDecisions()),
	})
}

func (h *Handlers) DecideClaim(c *gin.Context) {
	if _, resolved := caller.ContextFrom(c.Request.Context()); !resolved {
		httpx.Fail(c, httpx.CodeUnauthenticated)
		return
	}

	idempotencyKey, present := httpx.IdempotencyKeyFrom(c.Request.Context())
	if !present {
		httpx.Fail(c, httpx.CodeIdempotencyKeyRequired)
		return
	}

	var body decideRequest
	if err := c.ShouldBindJSON(&body); err != nil {
		httpx.Fail(c, httpx.CodeValidation)
		return
	}

	outcome, known := decisionOutcomeByName[body.Outcome]
	if !known {
		httpx.Fail(c, httpx.CodeValidation, httpx.FieldError{
			Field: "outcome", Code: "UNSUPPORTED_VALUE",
		})
		return
	}

	decided, err := h.Sustainability.DecideClaim(c.Request.Context(), idempotencyKey,
		&sustainabilityv1.DecideClaimRequest{
			ClaimId:        c.Param("claimId"),
			Outcome:        outcome,
			ApprovedAmount: body.ApprovedAmount,
			Reason:         body.Reason,
			IdempotencyKey: idempotencyKey,
		})
	if err != nil {
		h.failClaim(c, err)
		return
	}

	httpx.Data(c, http.StatusOK, map[string]any{
		"claim":     toClaimResponse(decided.GetClaim()),
		"decisions": toDecisionResponses(decided.GetDecisions()),
	})
}
